### ISSUE TITLE
Is pre-downloading images required?

### ISSUE THREAD

**Issue by dhood** [canonical/colcon-in-container#13]
Hi, I'd like to try this out, it looks useful! But on my first attempt I get:

```
colcon --log-level=debug build-in-container --ros-distro rolling
DEBUG:colcon:Parsed command line arguments: Namespace(log_base=None, log_level=10, verb_name='build-in-container', ros_distro='rolling', colcon_build_args='', debug=False, shell_after=False, provider='lxd', ignore_user_meta=False, metas=['./colcon.meta'], base_paths=['.'], packages_ignore=None, packages_ignore_regex=None, paths=None, packages_up_to=None, packages_up_to_regex=None, packages_above=None, packages_above_and_dependencies=None, packages_above_depth=None, packages_select_by_dep=None, packages_skip_by_dep=None, packages_skip_up_to=None, packages_select_build_failed=False, packages_skip_build_finished=False, packages_select_test_failures=False, packages_skip_test_passed=False, packages_select=None, packages_skip=None, packages_select_regex=None, packages_skip_regex=None, packages_start=None, packages_end=None, allow_overriding=[], mixin_files=None, mixin=None, verb_parser=<colcon_mixin.mixin.mixin_argument.MixinArgumentDecorator object at 0x7eb0f9e81270>, verb_extension=<colcon_in_container.verb.build_in_container.BuildInContainerVerb object at 0x7eb0fa691a20>, main=<bound method BuildInContainerVerb.main of <colcon_in_container.verb.build_in_container.BuildInContainerVerb object at 0x7eb0fa691a20>>, mixin_verb=('build-in-container',))
INFO:colcon.colcon_core.location:Using log path 'log/build-in-container_2024-05-07_19-54-23'
[0.118s] INFO:colcon.colcon-in-container:Downloading the image then creating the LXD instance
[0.130s] ERROR:colcon:colcon build-in-container: Failed getting remote image info: Failed getting image: The requested image couldn't be found for fingerprint "ubuntu/jammy/cloud/amd64"
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/colcon_core/command.py", line 533, in verb_main
    rc = context.args.main(context=context)
  File "/home/dhood/.local/lib/python3.10/site-packages/colcon_in_container/verb/build_in_container.py", line 99, in main
    self.provider = ProviderFactory.create(context.args.provider,
  File "/home/dhood/.local/lib/python3.10/site-packages/colcon_in_container/providers/provider_factory.py", line 47, in create
    return provider(ros_distro)  # type: ignore
  File "/home/dhood/.local/lib/python3.10/site-packages/colcon_in_container/providers/lxd.py", line 89, in __init__
    self.instance = self.lxd_client.instances.create(config, wait=True)
  File "/home/dhood/.local/lib/python3.10/site-packages/pylxd/models/instance.py", line 343, in create
    client.operations.wait_for_operation(response.json()["operation"])
  File "/home/dhood/.local/lib/python3.10/site-packages/pylxd/models/operation.py", line 57, in wait_for_operation
    operation.wait()
  File "/home/dhood/.local/lib/python3.10/site-packages/pylxd/models/operation.py", line 94, in wait
    response = self._client.api.operations[self.id].wait.get()
  File "/home/dhood/.local/lib/python3.10/site-packages/pylxd/client.py", line 207, in get
    self._assert_response(
  File "/home/dhood/.local/lib/python3.10/site-packages/pylxd/client.py", line 178, in _assert_response
    raise exceptions.LXDAPIException(response)
pylxd.exceptions.LXDAPIException: Failed getting remote image info: Failed getting image: The requested image couldn't be found for fingerprint "ubuntu/jammy/cloud/amd64"

```

is there some lxd image I should download ahead of time, or something?

**Comment by artivis** [comment]
Hello @dhood :wave:,

Thanks for your interest in this project! There is no need to pre-download images, this will happen automatically. The remote URL is broken at the moment but this should be fixed on the `dev/jazzy` branch if you wanna give it another go.

**Comment by dhood** [comment]
hey you!! :grin: 
nice tip - it helped it work for humble. Any tips up your sleeve for rolling? happened 3 times

```
dhood@dhood-t14:~/roscontrol_ws$ colcon --log-level=debug build-in-container --ros-distro rolling
DEBUG:colcon:Parsed command line arguments: Namespace(log_base=None, log_level=10, verb_name='build-in-container', ros_distro='rolling', colcon_build_args='', debug=False, shell_after=False, provider='lxd', ignore_user_meta=False, metas=['./colcon.meta'], base_paths=['.'], packages_ignore=None, packages_ignore_regex=None, paths=None, packages_up_to=None, packages_up_to_regex=None, packages_above=None, packages_above_and_dependencies=None, packages_above_depth=None, packages_select_by_dep=None, packages_skip_by_dep=None, packages_skip_up_to=None, packages_select_build_failed=False, packages_skip_build_finished=False, packages_select_test_failures=False, packages_skip_test_passed=False, packages_select=None, packages_skip=None, packages_select_regex=None, packages_skip_regex=None, packages_start=None, packages_end=None, allow_overriding=[], mixin_files=None, mixin=None, verb_parser=<colcon_mixin.mixin.mixin_argument.MixinArgumentDecorator object at 0x721f51a29990>, verb_extension=<colcon_in_container.verb.build_in_container.BuildInContainerVerb object at 0x721f5229da20>, main=<bound method BuildInContainerVerb.main of <colcon_in_container.verb.build_in_container.BuildInContainerVerb object at 0x721f5229da20>>, mixin_verb=('build-in-container',))
INFO:colcon.colcon_core.location:Using log path 'log/build-in-container_2024-05-08_16-54-47'
[3.044s] INFO:colcon.colcon-in-container:Downloading the image then creating the LXD instance
[5.739s] INFO:colcon.colcon-in-container:Waiting for ROS 2 to be installed
[6.077s] DEBUG:colcon.colcon-in-container.instance:.
...
195.423s] DEBUG:colcon.colcon-in-container.instance:.
[195.689s] DEBUG:colcon.colcon-in-container.instance:.
[195.956s] DEBUG:colcon.colcon-in-container.instance:.
[196.215s] DEBUG:colcon.colcon-in-container.instance:.
[196.471s] DEBUG:colcon.colcon-in-container.instance:
status: error

[197.010s] ERROR:colcon.colcon-in-container:Failed to run cloud-init with error: 1.
```

**Comment by artivis** [comment]
I'm afraid rolling is gonna be broken until we support jazzy (and thus 24.04) as rolling transitions to 24.04 :s

**Comment by dhood** [comment]
gah, fair enough. I'm in an interesting situation where I want to test rolling branches, but can be on 22.04 still, doesn't have to be 24.04. since ros2 packages for rolling _have_ been released to 22.04. But I appreciate that's a bit niche use case! Thanks for the suggestions, in any case
